Mathlib/FieldTheory/PurelyInseparable/Exponent.lean

```diff
@@ -0,0 +1,193 @@
+  for all `a ∈ L`.
+- `IsPurelyInseparable.exponent`: the exponent of a purely inseparable field extension.
+- `IsPurelyInseparable.elemExponent`: the exponent of an element of a purely inseparable
+  field extension, that is the smallest natural number `e` such that `a ^ ringExpChar K ^ e ∈ K`.
+
+## Tags
+
+purely inseparable
+
+-/
+
+namespace IsPurelyInseparable
+
+variable (K L : Type*)
+
+section Ring
+
+variable [CommRing K] [Ring L] [Algebra K L]
+
+/-- A predicate class on a ring extension saying that there is a natural number `e`
+such that `a ^ ringExpChar K ^ e ∈ K` for all `a ∈ L`. -/
+class HasExponent : Prop where
+  has_exponent : ∃ e, ∀ a, a ^ ringExpChar K ^ e ∈ (algebraMap K L).range
+
+/-- Version of `HasExponent.has_exponent` using `[ExpChar K p]`. -/
+theorem HasExponent.has_exponent' (p : ℕ) [ExpChar K p] [HasExponent K L] :
+    ∃ e, ∀ a, a ^ p ^ e ∈ (algebraMap K L).range :=
+  ringExpChar.eq K p ▸ has_exponent
+
+open scoped Classical in
+/-- The *exponent* of a purely inseparable extension is the smallest
+natural number `e` such that `a ^ ringExpChar K ^ e ∈ K` for all `a ∈ L`. -/
+noncomputable def exponent [HasExponent K L] : ℕ :=
+  Nat.find ‹HasExponent K L›.has_exponent
+
+variable {L}
+
+open Classical in
+theorem exponent_def [HasExponent K L] (a : L) :
+    a ^ ringExpChar K ^ exponent K L ∈ (algebraMap K L).range :=
+  Nat.find_spec ‹HasExponent K L›.has_exponent a
+
+/-- Version with `[ExpChar K p]`. -/
+theorem exponent_def' [HasExponent K L] (p : ℕ) [ExpChar K p] (a : L) :
+    a ^ p ^ exponent K L ∈ (algebraMap K L).range :=
+  ringExpChar.eq K p ▸ exponent_def K a
+
+variable {K}
+
+open Classical in
+theorem exponent_min [HasExponent K L] {e : ℕ} (h : e < exponent K L) :
+    ∃ a, a ^ ringExpChar K ^ e ∉ (algebraMap K L).range :=
+  not_forall.mp <| Nat.find_min ‹HasExponent K L›.has_exponent h
+
+/-- Version with `[ExpChar K p]`. -/
+theorem exponent_min' [HasExponent K L] (p : ℕ) [ExpChar K p] {e : ℕ} (h : e < exponent K L) :
+    ∃ a, a ^ p ^ e ∉ (algebraMap K L).range :=
+  ringExpChar.eq K p ▸ exponent_min h
+
+end Ring
+
+section IsDomain
+
+variable [Field K] [Ring L] [IsDomain L] [Algebra K L]
+
+instance [HasExponent K L] : IsPurelyInseparable K L :=
+  let ⟨n, h⟩ := ‹HasExponent K L›.has_exponent
+  (isPurelyInseparable_iff_pow_mem K (ringExpChar K)).mpr fun x ↦ ⟨n, h x⟩
+
+end IsDomain
+
+section Field
+
+open Polynomial
+
+variable [Field K] [Field L] [Algebra K L] [IsPurelyInseparable K L]
+variable {L}
+
+open Classical in
+/-- The exponent of an element `a ∈ L` of a purely inseparable field extension `L / K`
+is the smallest natural number `e` such that `a ^ ringExpChar K ^ e ∈ K`. -/
+noncomputable def elemExponent (a : L) : ℕ :=
+  Nat.find <| minpoly_eq_X_pow_sub_C K (ringExpChar K) a
+
+open Classical in
+variable {K} in
+theorem elemExponent_eq_zero_of_mem_range {a : L} (h : a ∈ (algebraMap K L).range) :
+    elemExponent K a = 0 := by
+  apply (Nat.find_eq_zero _).mpr
+  simp
```

Comment:
I think non-terminal `simp` is not allowed in mathlib code. Could you try replace it by the output of `simp?` which looks like `simp only [XXX]` ?